packages/cli/src/cmds/ask/ask.ts

```diff
@@ -0,0 +1,307 @@
+          eventMatchesByDiagramId.set(diagramId, [event]);
+        }
+        */
+      });
+    }
+
+    // Collect all diagrams that match all keywords
+    const matchingDiagrams = new Set<string>();
+    for (const diagramId of matchesForKeyword[0]) {
+      let isMatch = true;
+      for (const matches of matchesForKeyword) {
+        if (!matches.has(diagramId)) {
+          isMatch = false;
+          break;
+        }
+      }
+      if (isMatch) matchingDiagrams.add(diagramId);
+    }
+
+    return Array.from(matchingDiagrams).map((diagramId) => ({ diagramId }));
+    // return Array.from(eventMatchesByDiagramId.entries()).map(([diagramId, matchingEvents]) => ({
+    //   diagramId,
+    //   matchingEvents,
+    // }));
+  }
+}
+
+export const handler = async (argv: any) => {
+  verbose(argv.verbose);
+  handleWorkingDirectory(argv.directory);
+  const appmapDir = await locateAppMapDir(argv.appmapDir);
+
+  const ask = new Ask(appmapDir);
+
+  function showPlan(paramStr: string) {
+    const params = JSON.parse(paramStr) as { plan: string };
+    warn(`Plan:\n${params.plan}`);
+  }
+
+  async function searchDiagrams(paramStr: string): Promise<SearchDiagramResult[]> {
+    const params = JSON.parse(paramStr) as SearchDiagramsParam;
+    let { keywords } = params;
+    warn(`Searching for diagrams matching "${keywords}"`);
+    const eventMatches = await ask.searchDiagrams(keywords);
+    if (!eventMatches) return [];
+
+    return eventMatches;
+  }
+
+  async function getDiagramDetails(paramStr: string): Promise<DiagramDetailsResult[]> {
+    const params = JSON.parse(paramStr) as DiagramDetailsParam;
+    const { diagramIds } = params;
+    warn(`Getting details for diagram ${diagramIds}`);
+    const result = new Array<DiagramDetailsResult>();
+    for (const diagramId of diagramIds) {
+      const metadata = JSON.parse(
+        await readFile(join(diagramId, 'metadata.json'), 'utf-8')
+      ) as Metadata;
+      delete metadata['git'];
+
+      warn(`Building sequence diagram for AppMap ${diagramId}`);
+      const appmapFile = [diagramId, 'appmap.json'].join('.');
+      const prunedAppMap = buildAppMap()
+        .source(await readFile(appmapFile, 'utf-8'))
+        .prune(10 * 1000 * 1000)
+        .build();
+      const filter = new AppMapFilter();
+      if (metadata.language?.name !== 'java') filter.declutter.hideExternalPaths.on = true;
+
+      filter.declutter.limitRootEvents.on = true;
+      const filteredAppMap = filter.filter(prunedAppMap, []);
+      const specification = Specification.build(filteredAppMap, { loops: true });
+      const diagram = buildDiagram(appmapFile, filteredAppMap, specification);
+
+      const diagramText: string[] = [];
+      const maxDepth = 2;
+      const collectAction = (action: Action, depth = 0) => {
+        if (depth > maxDepth) return;
+
+        const indent = '  '.repeat(depth);
+        diagramText.push(`${indent}${nodeName(action)}(events:${action.eventIds.join(', ')}})`);
```

Comment:
FYI There's a stray curly brace at the end of this string. Not sure how that may affect accuracy.